src/background/http-instrument.ts

```diff
@@ -1,687 +1,693 @@
+    const isFrameLoad = details.type === "sub_frame";
+    update.is_full_page = boolToInt(isFullPageLoad);
+    update.is_frame_load = boolToInt(isFrameLoad);
+
+    // Grab the triggering and loading Principals
+    let triggeringOrigin;
+    let loadingOrigin;
+    if (details.originUrl) {
+      const parsedOriginUrl = new URL(details.originUrl);
+      triggeringOrigin = parsedOriginUrl.origin;
     }
-  } catch (anError) {
-    // Exceptions expected for channels triggered or loading in a
-    // NullPrincipal or SystemPrincipal. They are also expected for favicon
-    // loads, which we attempt to filter. Depending on the naming, some favicons
-    // may continue to lead to error logs.
-    if (
-      update.triggering_origin !== "[System Principal]" &&
-      update.triggering_origin !== undefined &&
-      update.loading_origin !== "[System Principal]" &&
-      update.loading_origin !== undefined &&
-      !update.url.endsWith("ico")
-    ) {
-      loggingDB.logError(
-        "Error while retrieving additional channel information for URL: " +
+    if (details.documentUrl) {
+      const parsedDocumentUrl = new URL(details.documentUrl);
+      loadingOrigin = parsedDocumentUrl.origin;
+    }
+    update.triggering_origin = escapeString(triggeringOrigin);
+    update.loading_origin = escapeString(loadingOrigin);
+
+    // loadingDocument's href
+    // The loadingDocument is the document the element resides, regardless of
+    // how the load was triggered.
+    const loadingHref = details.documentUrl;
+    update.loading_href = escapeString(loadingHref);
+
+    // resourceType of the requesting node. This is set by the type of
+    // node making the request (i.e. an <img src=...> node will set to type "image").
+    // Documentation:
+    // https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/webRequest/ResourceType
+    update.resource_type = details.type;
+
+    /*
+    // TODO: Refactor to corresponding webext logic or discard
+    const ThirdPartyUtil = Cc["@mozilla.org/thirdpartyutil;1"].getService(
+                           Ci.mozIThirdPartyUtil);
+    // Do third-party checks
+    // These specific checks are done because it's what's used in Tracking Protection
+    // See: http://searchfox.org/mozilla-central/source/netwerk/base/nsChannelClassifier.cpp#107
+    try {
+      const isThirdPartyChannel = ThirdPartyUtil.isThirdPartyChannel(details);
+      const topWindow = ThirdPartyUtil.getTopWindowForChannel(details);
+      const topURI = ThirdPartyUtil.getURIFromWindow(topWindow);
+      if (topURI) {
+        const topUrl = topURI.spec;
+        const channelURI = details.URI;
+        const isThirdPartyToTopWindow = ThirdPartyUtil.isThirdPartyURI(
+          channelURI,
+          topURI,
+        );
+        update.is_third_party_to_top_window = isThirdPartyToTopWindow;
+        update.is_third_party_channel = isThirdPartyChannel;
+        update.top_level_url = escapeString(topUrl);
```

Comment:
`top_level_url` is super important to have, and would let us compute the other two properties (i.e., `is_third_party_to_top_window` and `is_third_party_channel`) by comparing to `url` and `document_url` if there isn't a direct way to grab them.

We can probably derive this from `tabId`. Note that we're probably going to run into the issues identified by @ghostwords in [Bug 1499523](https://bugzilla.mozilla.org/show_bug.cgi?id=1499523) and [Bug 1470537](https://bugzilla.mozilla.org/show_bug.cgi?id=1470537#c2), and you may be able to reference the Privacy Badger codebase for an approach to computing this.

Reply:
Hello! I've got an in-progress PR to work around at least some of the issues related to relying on `chrome.webRequest` "main_frame" requests to identify which URL belongs to which tab: https://github.com/EFForg/privacybadger/pull/2198. It's on pause until we finish and release a higher-priority set of fixes.

Reply:
@ghostwords I'd love your feedback in relation to this in https://github.com/mozilla/openwpm-webext-instrumentation/issues/9

Reply:
See 1e61c1f (fix: Restore HttpRequest.top_level_url (using the url of the requesting tab)) for the naive way of implementing this for the time being.